### DISCUSSION TITLE
Configurable path to App

### DISCUSSION THREAD

**Discussion by danieljwestman** [directus/directus#2988] [Feature Requests]
It would be a great enhancement if we where able to choose our own path to the App. As for now it's hardcoded to `/admin`. In my humble opinion thats not suitable for projects where many users with different kinds of roles has access to the App (not only _admins_). The Directus App isn't just an admin interface - it's more. A more suitable path would be `/app` - but even better if it's configurable in the `.env` file!

There are 5 files to modify (which I've tried and successfully achieved with the above).

**/api/src/app.ts on line 102**
```
		app.get('/', (req, res) => res.redirect(`./admin/`));
		app.get('/admin', (req, res) => res.send(html));
		app.use('/admin', express.static(path.join(adminPath, '..')));
		app.use('/admin/*', (req, res) => {
			res.send(html);
		});
```

**/api/src/services/users.ts on line 93 & 134**
```
		const acceptURL = env.PUBLIC_URL + '/admin/accept-invite?token=' + token;
...
		const acceptURL = env.PUBLIC_URL + '/admin/reset-password?token=' + token;
```

**/app/vue.config.js on line 8 & 18**
```
	publicPath: '/admin/',

	devServer: {
		allowedHosts: ['localhost', '.gitpod.io'],
		port: 8080,
		public: '127.0.0.1:8080',
		proxy: {
			'/': {
				target: process.env.API_URL ? process.env.API_URL : 'http://localhost:8055/',
				changeOrigin: true,
				bypass: (req) => (req.url.startsWith('/admin') ? req.url : null),
			},
		},
		progress: false,
	},
```

**/app/src/main.ts on line 3**
```
__webpack_public_path__ = getRootPath() + 'admin/';
```

**/app/src/router.ts on line 73**
```
const router = new VueRouter({
	mode: 'history',
	base: getRootPath() + 'admin/',
	routes: defaultRoutes,
});

export function replaceRoutes(routeFilter: (routes: RouteConfig[]) => RouteConfig[]): void {
	const newRoutes = routeFilter([...defaultRoutes]);
	const newRouter = new VueRouter({
		mode: 'history',
		base: getRootPath() + 'admin/',
		routes: newRoutes,
	});

	// @ts-ignore - Matcher is not officially part of the public API (https://github.com/vuejs/vue-router/issues/2844#issuecomment-509529927)
	router.matcher = newRouter.matcher;
}
```

**Comment by rijkvanzanten** [comment]
The `publicPath: '/admin/',` in `vue.config.js` is the biggest issue. That's controlling the build settings in build time, which can't be configured in regular installs. We'll have to do some research what that `publicPath` settings does in `@vue/cli-service` and make sure we patch whatever it sets in runtime so we can configure the path in an ENV var

**Reply by danieljwestman** [reply]
Thats only in dev mode, right? I managed to get it working by running `APP_PATH=app npm run dev` and also setting it in the .env file.

**Reply by Fusseldieb** [reply]
https://cli.vuejs.org/config/#publicpath

> The value can also be set to an empty string ('') or a relative path (./) so that all assets are linked using relative paths.

---

> Thats only in dev mode, right? I managed to get it working by running `APP_PATH=app npm run dev` and also setting it in the .env file.

From the link above:

> This value is also respected during development. If you want your dev server to be served at root instead, you can use a conditional value:
> ```
>module.exports = {
>  publicPath: process.env.NODE_ENV === 'production'
>    ? '/production-sub-path/'
>    : '/'
>}
>```

I hope I'm not understanding things in this discussion wrong and talking about nonsense.

EDIT: Should be possible to use template literals, like:

 ```
module.exports = {
  publicPath: process.env.NODE_ENV === 'production'
    ? `/${process.env.APP_PATH}/`
    : '/admin'
}
```

Or just:

 ```
module.exports = {
  publicPath: `/${process.env.APP_PATH}/`
}
```

**Reply by danieljwestman** [reply]
Yep, thats how I did it. 
`publicPath: /${process.env.APP_PATH}/`

But it doesn’t seem to get vars from the env file in this case, therefore the special npm run command. Should be a better way...

**Reply by Fusseldieb** [reply]
I think I need to make the modifications for myself to see what happens. I am afraid to give unhelpful tips if I don't 🤣

**Reply by danieljwestman** [reply]
Go ahead. Just replace all `admin` with the env variable at the lines in the snippets I’ve attached in the post above.

**Reply by rijkvanzanten** [reply]
This is the key bit:

`so that all assets are linked using relative paths.`

That means that the final dist/index.html contains hardcoded `/admin/script.js` in the src-es of the styles/scripts. I already have a bit of script in the API that prefixes these paths with the APIs public URL, so it should be possible to replace the `/admin` bit in that same logic on the fly.

From the dist index.html of the app:

![CleanShot 2020-11-12 at 14 16 55](https://user-images.githubusercontent.com/9141017/98985591-c07bf700-24f1-11eb-95e0-a1e13c34a86d.png)


(these paths need to stay absolute, otherwise refreshes break (as it'll try loading /admin/collections/articles/style.css)

**Reply by Fusseldieb** [reply]
Thank you for the heads-up @rijkvanzanten and @danieljwestman . Now it makes more sense. I'll definitively try this and report back.

**Reply by danieljwestman** [reply]
@Fusseldieb Looked into this?

**Reply by Fusseldieb** [reply]
I will try it today, @danieljwestman ! :)

**Reply by danieljwestman** [reply]
Nice, looking forward to see 👍🏻

**Comment by ernst86913** [comment]
Hi everyone - where you able to verify it? This still would be a major improvement.....  Thanks, Roman

**Reply by paescuj** [reply]
AFAIK this isn't possible yet due to the difficulties described above.

**Reply by rijkvanzanten** [reply]
This feature request hasn't really gotten any traction/upvotes in the community, so it's not something we're actively researching at the moment.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
